File: algorithms/0876-middle-of-the-linked-list/Solution.java
// Middle of the Linked List - fast/slow pointers.
// Time O(n), Space O(1).

class ListNode {
    int val;
    ListNode next;
    ListNode(int val) { this.val = val; }
}

class Solution {
    public ListNode middleNode(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }
}

File: algorithms/0876-middle-of-the-linked-list/Solution.kt
// Middle of the Linked List - fast/slow pointers.
// Time O(n), Space O(1).

class ListNode(var `val`: Int, var next: ListNode? = null)

class Solution {
    fun middleNode(head: ListNode?): ListNode? {
        var slow = head
        var fast = head
        while (fast?.next != null) {
            slow = slow?.next
            fast = fast.next!!.next
        }
        return slow
    }
}
